### PR TITLE
Properly create code unit systems

### PR DESCRIPTION
**Pull request by jzuhone**
The current behavior of yt with respect to "code" unit systems to to permit switching a `YTArray` to a code unit system via something like:

```python
ds = yt.load("my_dataset")
sp = ds.sphere("c", (100.0, "kpc"))
sp["density"].in_base(ds)
```

which is a bit clunky and not terribly symmetric with the other ways of switching to different unit systems, e.g. `in_base("galactic")`. In the current implementation, the code unit systems in the `unit_system_registry` are indexed by the string representation of the dataset.

With respect to this latter point, this is actually a problem, because for stream datasets this representation is actually the same for every dataset, so one would end up overwriting the unit system in the registry every time one opened a new one. The solution is for every dataset to have a unique identifier.

So that each dataset has a unique hash within the `unit_system_registry` dict, I've stolen the `fnv_hash` from `selection_routines` and placed it in a Cython file under `yt/utilities/lib`, and used it for this purpose.

The other benefit to this technique is that one can now use `in_base("code")` for `YTArray`s with code units and it will work, in the same way as the other unit systems.

I need to edit the docs, but otherwise this is ready for code review.

UPDATE: I changed the warning message that occurs when someone sets `units_override` to something less scary-sounding.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
It looks like this somehow affects the athena answer tests and also breaks the GDF frontend. Let me know if you want a hand fixing the tests.

File: yt/frontends/gdf/data_structures.py [hunk 1]
```diff
@@ -222,6 +224,12 @@ def _set_code_unit_attributes(self):
                 # comparisons are insufficient
                 unit = Unit(unit, registry=self.unit_registry)
                 if unit_name.endswith('_unit') and unit.dimensions is sympy_one:
+                    # Catch code units and if they are dimensionless,
```
Comment (jzuhone) [inline yt/frontends/gdf/data_structures.py]: @ngoldbaum and @Xarthisius should have a look at the changes I made here.

**Comment by jzuhone** [comment]
Ok, so I have figured out everything but the Athena++ failures, and I'm a bit stuck. I can't reproduce those failures locally, actually.

**Comment by jzuhone** [comment]
Never mind, I think I figured this out.

**Comment by jzuhone** [comment]
Ok, I am actually convinced now that the Athena++ answers are now correct as a result of this PR--it looks like the magnetic field calculation hasn't been right for a while. I have changed the answer number accordingly.

File: yt/utilities/lib/fnv_hash.pxd [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+
+
+import numpy as np
+cimport numpy as np
+
+cdef np.int64_t _fnv_hash(unsigned char[:] octets)
```
Comment (ngoldbaum) [inline yt/utilities/lib/fnv_hash.pxd]: Minor style nit, maybe call this `c_fnv_hash`? I think that there's a `fnv_hash` exposed in python and `_fnv_hash` exposed in cython is a bit confusing. Took me a minute or two to figure out anyway.

File: yt/utilities/lib/fnv_hash.pyx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    # https://bitbucket.org/yt_analysis/yt/issues/1052/field-access-tests-fail-under-python3
+    # FNV hash cf. http://www.isthe.com/chongo/tech/comp/fnv/index.html
+    cdef np.int64_t hash_val = 2166136261
+    cdef char octet
+    for octet in octets:
+        hash_val = hash_val ^ octet
+        hash_val = hash_val * 16777619
+    return hash_val
+
+def fnv_hash(octets):
```
Comment (ngoldbaum) [inline yt/utilities/lib/fnv_hash.pyx]: Maybe a docstring? Documenting that `octets` needs to be a bytestring would be useful.

File: yt/units/unit_registry.py [hunk 4]
```diff
@@ -53,6 +54,17 @@ def __getitem__(self, key):
     def __contains__(self, item):
         return item in self.lut
 
+    _unit_system_id = None
+    @property
+    def unit_system_id(self):
```
Comment (ngoldbaum) [inline yt/units/unit_registry.py]: Docstring?

File: yt/frontends/gdf/data_structures.py (hunk 1 above)
Comment (ngoldbaum) [inline yt/frontends/gdf/data_structures.py]: Looks fine to me

File: yt/frontends/athena_pp/data_structures.py [hunk 5]
```diff
@@ -278,16 +278,9 @@ def _set_code_unit_attributes(self):
             mylog.warning("Assuming 1.0 = 1.0 %s", cgs)
             setattr(self, "%s_unit" % unit, self.quan(1.0, cgs))
 
-    def set_code_units(self):
-        super(AthenaPPDataset, self).set_code_units()
```
Comment (ngoldbaum) [inline yt/frontends/athena_pp/data_structures.py]: Why aren't we calling `set_code_units` on the superclass? For example, will `_set_code_unit_attributes` still be called without this?

Comment (jzuhone) [inline yt/frontends/athena_pp/data_structures.py]: I'm copying here the logic that you implemented for the Athena frontend, [see here](https://bitbucket.org/yt_analysis/yt/pull-requests/2213/mks-code-unit-system-fixes-closes-1229/diff). This is the real fix for the Athena++ bug. `_set_code_unit_attributes` is called. It's unnecessary to subclass `set_code_units` here.

**Comment by jzuhone** [comment]
@ngoldbaum could you elaborate on what part of this exactly you'd like to see another regression test for?

File: yt/frontends/athena_pp/data_structures.py (hunk 5 above)
Comment (ngoldbaum) [inline yt/frontends/athena_pp/data_structures.py]: Ah, sorry I misread. I thought the function definition wasn't being deleted.

**Comment by ngoldbaum** [comment]
Weren't you originally running into issues comparing two stream datasets? Maybe something along those lines. You could create two different stream datasets and verify that they have different code unit systems.

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Pull request merged**